proxy/core/acceptor/threadless.py

```diff
@@ -93,44 +97,35 @@ def selected_events(self) -> Generator[
                 # asynchronous nature.  Hence, threadless will handle
                 # ValueError exceptions raised by selector.register
                 # for invalid fd.
-                self.selector.register(fd, worker_events[fd])
+                self.selector.register(fd, worker_events[fd], data=work_id)
         ev = self.selector.select(timeout=DEFAULT_SELECTOR_SELECT_TIMEOUT)
-        readables = []
-        writables = []
+        # Keys are work_id and values are 2-tuple indicating readables & writables
+        work_by_ids: Dict[int, Tuple[Readables, Writables]] = {}
         for key, mask in ev:
+            if key.data not in work_by_ids:
+                work_by_ids[key.data] = ([], [])
             if mask & selectors.EVENT_READ:
-                readables.append(key.fileobj)
+                work_by_ids[key.data][0].append(key.fileobj)
             if mask & selectors.EVENT_WRITE:
-                writables.append(key.fileobj)
-        yield (readables, writables)
-        for fd in events:
-            self.selector.unregister(fd)
-
-    async def handle_events(
-            self, fileno: int,
-            readables: Readables,
-            writables: Writables
-    ) -> bool:
-        return self.works[fileno].handle_events(readables, writables)
+                work_by_ids[key.data][1].append(key.fileobj)
+        return (events, work_by_ids)
 
     # TODO: Use correct future typing annotations
     async def wait_for_tasks(
-            self, tasks: Dict[int, Any]
+            self,
+            tasks: Set['asyncio.Task[bool]'],
     ) -> None:
-        for work_id in tasks:
-            # TODO: Resolving one handle_events here can block
-            # resolution of other tasks.  This can happen when handle_events
-            # is slow.
-            #
-            # Instead of sequential await, a better option would be to await on
-            # list of async handle_events.  This will allow all handlers to run
-            # concurrently without blocking each other.
-            try:
-                teardown = await asyncio.wait_for(tasks[work_id], DEFAULT_TIMEOUT)
-                if teardown:
-                    self.cleanup(work_id)
-            except asyncio.TimeoutError:
-                self.cleanup(work_id)
+        assert len(tasks) > 0
+        finished, self.unfinished = await asyncio.wait(
+            tasks,
+            timeout=(self.flags.num_workers / 2) *
+            DEFAULT_SELECTOR_SELECT_TIMEOUT,
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+        for task in finished:
+            if task.result():
+                self.cleanup(task._work_id)     # type: ignore
```

Comment:
Instead of ignoring all possible type errors, put the exact error code that you're seeing to prevent accidentally shadowing more problems that may arise in the future:
```suggestion
                self.cleanup(task._work_id)     # type: ignore[a-specific-error]
```

Reply:
Refactored code and this line is now within `Threadless`, applied there manually.

Reply:
`# type: ignore[attr-defined]` works